runtime/src/accounts_background_service/pending_snapshot_packages.rs

```diff
@@ -207,60 +285,104 @@ mod tests {
+            fastboot: None,
         };
 
         // pushing an older incremental should panic
         pending_snapshot_packages.push(new_incr(slot - 1, base));
     }
 
+    #[test]
+    #[should_panic(expected = "fastboot snapshot package must be newer than pending package")]
+    fn test_push_older_fastboot() {
+        let slot = 100;
+        let mut pending_snapshot_packages = PendingSnapshotPackages {
+            full: None,
+            incremental: None,
+            fastboot: Some(new_fastboot(slot)),
+        };
+
+        // pushing an older incremental should panic
+        pending_snapshot_packages.push(new_fastboot(slot - 1));
+    }
+
     #[test]
     fn test_pop() {
         let mut pending_snapshot_packages = PendingSnapshotPackages::default();
 
-        // ensure we can call pop when there are no pending packages
-        assert!(pending_snapshot_packages.pop().is_none());
+        let oldest_slot = 100;
+        let middle_slot = 200;
+        let newest_slot = 300;
 
-        // ensure pop returns full when there's only a full
-        let slot = 100;
-        pending_snapshot_packages.full = Some(new_full(slot));
-        pending_snapshot_packages.incremental = None;
-        let snapshot_package = pending_snapshot_packages.pop().unwrap();
-        assert!(snapshot_package.snapshot_kind.is_full_snapshot());
-        assert_eq!(snapshot_package.slot, slot);
+        let slots = vec![
+            None,
+            Some(oldest_slot),
+            Some(middle_slot),
+            Some(newest_slot),
+        ];
+        let mut scenarios = Vec::new();
+
+        // Generate all combinations of full, incremental, and fastboot slots
+        for &full in &slots {
+            for &incr in &slots {
+                for &fastboot in &slots {
+                    scenarios.push((full, incr, fastboot));
+                }
+            }
+        }
+
+        for (full, incr, fastboot) in scenarios {
+            pending_snapshot_packages.full = full.map(new_full);
+            pending_snapshot_packages.incremental = incr.map(|slot| new_incr(slot, slot));
+            pending_snapshot_packages.fastboot = fastboot.map(new_fastboot);
+
+            // Pop full snapshot if it exists
+            if let Some(full_slot) = full {
+                assert_eq!(pending_snapshot_packages.pop().unwrap().slot, full_slot);
+            }
+
+            // Pop incremental snapshot if it exists and is newer than the full snapshot
+            if let Some(incr_slot) = incr {
+                if full.is_none() || incr_slot > full.unwrap() {
+                    assert_eq!(pending_snapshot_packages.pop().unwrap().slot, incr_slot);
+                }
+            }
+
+            // Pop fastboot snapshot if it exists and is newer than both full and incremental
+            if let Some(fastboot_slot) = fastboot {
+                if fastboot_slot > full.unwrap_or(0) && fastboot_slot > incr.unwrap_or(0) {
```

Comment:
...but when combined with this one, it would be nice to have them all using the same 'style' of checks.

```suggestion
                if full.is_none_or(|full_slot| fastboot_slot > full_slot) && incr.is_none_or(|incr_slot| fastboot_slot > incr_slot) {
```